Use ubuntu clang and fix a bug in run_as_original_user

aneeshdurg opened the pull request.

tylershunt commented:
I think a valid question to ask before merging is "Will anyone use this but Tyler?" It's not that much more complicated to just run `conan install -if $HOME/build --build=missing config` when you see the build fail because you're missing some library, but even to write this comment I had to look that up which is why I started looking into alt config paths in the first place.

The people most suited to answer this question are people like @witchel @bowu and simon (will edit with simon's github tag). Since they are people that have had to adapt to our tools relatively recently.

I'll collect an informal poll here to get the general temperature: 👍 if you would use this 👎 if you think it's needless cruft.

l-hoang commented:
As someone who hasn't tried out the conan build yet, if it makes my life easier I'm for it. (can't comment on how this would make my life easier since I'm unfamiliar with how conan works, but from a quick glance it seems like it makes it so I don't have do something if there's a dependency update).

bowu commented:
I think it's useful, especially because I had to check the document several times for just this command recently.

tylershunt commented:
That's enough support for me! I'll merge it.

aneeshdurg committed:
Use ubuntu clang and fix a bug in run_as_original_user

The clang-10 package is now availible in the default ubuntu repos and
has been rotated out of the llvm repos.

aneeshdurg commented:
`run_as_original_user` didn't actually the run command passed in previously and instead echoed the literal string `'$@'` without doing any argument expansion

ddn0 commented on scripts/setup_ubuntu.sh:
```diff
@@ -21,14 +21,12 @@ for arg in "$@"; do
+
   if [[ -n "${ORIGINAL_USER}" ]]; then
-    su - "${ORIGINAL_USER}" bash -c '"$@"' -- "$@"
+    su - "${ORIGINAL_USER}" bash -c "$*"
```
The code was written to support things like:
```
run_as_original_user foo pip install --upgrade pip<=20.3.4
```
as strictly stringifying would lead to `<` being interpreted as a pipe redirect.

aneeshdurg replied:
Maybe a cleaner solution here is to make a `requirements.txt`? Feels weird to have extra complexity we're not using here.

arthurp replied:
We could generate and use a requirements file in the script. I don't think we could include the requirement file in the repo because in at least one script the packages are selected based on the environment the script is executing in.

That said let's not make this `run_as_original_user` less safe. If it's failing let's fix it right. Bash is terrible, and not quoting things properly will just make things more confusing. When and how was `run_as_original_user` failing? It worked for me and I tested it a good amount, so it doesn't universally fail like your comment implies.

aneeshdurg replied:
actually it seems like the `$*` approach already handles the redirect stuff correctly. I still think a `requirements.txt` would be easier to deal with and would avoid this confusion entirely

aneeshdurg replied:
maybe it has to do with the fact the I'm using a weird shell and `su` executes the command in your login shell? I'm using `xonsh` as my login shell, so maybe the difference in variable expansion is somehow incompatible there.

arthurp replied:
I'm 99% sure it doesn't. This is what I get for a simple test:

```
$ function tttttt() { bash -c "$*"; }
$ tttttt echo "test<test"
bash: test: No such file or directory
```
(The `<` is being interpreted)

```
$ function tttttt() { bash -c '"$@"' -- "$@"; }
$ tttttt echo "test<test"
test<test
```
(The `<` is being echoed)

Yes. I think generating a requirements files (which we should NOT name `requirements.txt` because it's not actually package requirements in the standard sense) is great. But we should still keep the quote correct version of this function unless it can be removed entirely (which I don't think it can be).

arthurp replied:
> `xonsh` as my login shell

That could be part of the issue. But since the script is run with `/bin/bash` and the `su` command immediately calls `bash` I don't see how `xonsh` could be involved. At most it should be passing through the arguments without interpretation. But it's possible there is some login shell step I'm not aware of in `su`.

aneeshdurg replied:
ah that's weird, I tested it by making the setup script add a constraint on the package version and that seemed to work. If we do switch to generating the requirements file can't we just change this function to `install_pip_requirements_as_user` instead and then not worry about the redirection stuff? There's no uses of this function to do anything besides call `pip install`. 

Hmm, yeah the parameter expansion should be done by bash before the `su` call so my login shell shouldn't matter...

arthurp replied:
Yes that's true.

As a side note, I think you should probably debug the shell issue in your environment. If your environment is misinterpreting shell scripts you could have some really hard bugs down the road. And it seems unlikely that this is the only case where scripts will execute incorrectly because this is not a very complex set of substitutions semantically.

aneeshdurg replied:
ok, it was my login shell. running `su -s $(which bash) - ....` worked. I'll just add the `-s` flag here for now.

arthurp replied:
Perfect. Thanks for debugging this. And I'll try to remember this. Using `su` in a script isn't common, but it does happen and this issue could also affect people using `zsh`.

I'm still open to switching to use a file. But it's probably not worth that much engineering time.

ddn0 reviewed (commented):
For the "Use ubuntu clang" commit, can you add something to the commit message along the lines of: "...The clang-10 package has been rotated out of the LLVM package repos and recent versions of Ubuntu include it natively..."

cc @arthurp

aneeshdurg commented:
updated the description of the commit.

arthurp commented on scripts/setup_ubuntu.sh:
```diff
@@ -98,5 +93,5 @@ apt install -yq --allow-downgrades \
 # Toolchain variants
 if [[ -n "${SETUP_TOOLCHAIN_VARIANTS}" ]]; then
-  apt install -yq gcc-9 g++-9 clang-10 clang++-10
+  apt install -yq gcc-9 g++-9 clang-10
```
Why remove `clang++-10` and not `clang-10`?

aneeshdurg replied:
I was unable to install the package `clang++-10` locally, it seems like the binary `clang++-10` was provided by `clang-10`.

arthurp replied:
Ah. OK. Reasonable. Just make sure to test the script on 18.04, 20.04, and (as below) maybe 16.04. You can spin up fresh installs on AWS for testing if you need.

ddn0 replied:
It is probably easier to use docker to test different distros:

E.g.,
```
docker run --rm -it ubuntu:20.04
```

Anyway, these scripts aren't used for 16.04. That goes through the packaging code:

https://github.com/KatanaGraph/katana-enterprise/tree/master/deploy/package

`scripts/setup_ubuntu.sh` is just used by CI and `setup_dev_ubuntu.sh`

arthurp commented:
Your commit message says "The default ubuntu repos". Which release do you mean? We have to support several. Have they back ported clang-10?

aneeshdurg commented:
> Your commit message says "The default ubuntu repos". Which release do you mean? We have to support several. Have they back ported clang-10?

It's present in both the `bionic` and `focal` releases. 

https://packages.ubuntu.com/search?suite=bionic&section=all&arch=any&keywords=clang-10&searchon=names
https://packages.ubuntu.com/search?suite=focal&section=all&arch=any&keywords=clang-10&searchon=names

arthurp commented:
> It's present in both the `bionic` and `focal` releases.

Great! Sadly we still have some lingering need for 16.04 (Xenial), which we have not been able to kill yet. @bosconi, do you still need 16.04? If so we may not be able to remove this code unconditionally.

tylershunt commented:
> > It's present in both the `bionic` and `focal` releases.
> 
> Great! Sadly we still have some lingering need for 16.04 (Xenial), which we have not been able to kill yet. @bosconi, do you still need 16.04? If so we may not be able to remove this code unconditionally.

Yeah, being able to build on 16.04 is still a requirement for an active customer engagement.

ddn0 commented:
> Yeah, being able to build on 16.04 is still a requirement for an active customer engagement.

This is true, but we don't use `setup_ubuntu.sh` directly to do this, right?

aneeshdurg committed:
Merge branch 'master' of ssh://github.com/KatanaGraph/katana into HEAD

tylershunt commented:
> > Yeah, being able to build on 16.04 is still a requirement for an active customer engagement.
> 
> This is true, but we don't use `setup_ubuntu.sh` directly to do this, right?

Oh, I do when I need to build on GCP to debug. I can work around it though not a huge deal.

ddn0 commented:
@aneeshdurg my proposal is to add some distro checks like

https://github.com/KatanaGraph/katana/blob/master/scripts/setup_dev_ubuntu.sh#L9

in `setup_ubuntu.sh` as well. If 16.04, old stuff, otherwise, new stuff.

My preference would be to use the distro version numbers like:

https://github.com/KatanaGraph/katana-enterprise/blob/master/deploy/package/build-ubuntu#L21

because I can never remember the codenames for Ubuntu releases.

if you do, can you update `setup_dev_ubuntu.sh` to match?

arthurp commented:
> > > Yeah, being able to build on 16.04 is still a requirement for an active customer engagement.
> > 
> > 
> > This is true, but we don't use `setup_ubuntu.sh` directly to do this, right?
> 
> Oh, I do when I need to build on GCP to debug. I can work around it though not a huge deal.

Let's not "half-support" anything, please. Let's support 16.04 as much as is practical, or not support it at all. The half way state is just super confusing even internally. I answer "why can't I build" questions relatively often and they are often cases of people not reading all the details, so having to say "well we support 16.04 but only with this specific funny setup that is different from other OSs" seems like a recipe for confusion.

(This could include OS version checks as proposed by @ddn0. That's still support. And I second that we should always use the version numbers for Ubuntu releases.)

aneeshdurg committed:
Add xenial specific installation steps

aneeshdurg committed:
lint

aneeshdurg commented:
Added the check. I had it only check the major version since that seemed reasonable.

ddn0 reviewed (approved):
LGTM

arthurp commented on scripts/setup_dev_ubuntu.sh:
```diff
@@ -6,12 +6,17 @@
+RELEASE_MAJOR=$(lsb_release --release --short | cut -d . -f 1)
+RELEASE_NAME=$(lsb_release --codename | awk '{print $2}')
 
 if echo "${RELEASE}" | grep -q -e "${EXPECTED_RELEASE}" > /dev/null;
```
This test is wrong now since you removed `RELEASE`. It's a good idea to test the failing case not just the successful one.

I would recommend just using a `case` statement. My use of `grep` was a bad idea I think. Best to get rid of that.

aneeshdurg replied:
yikes!

Maybe I'll look into adding shellcheck in our lint process.

arthurp reviewed (commented):
I know I'm late, but I think there are bug that will be a problem.

I don't think checking the major version is enough. The reason is that it implies we are supporting 18.10 which we are not. We only support the LTS releases in those years. It's no harder to check the full version. Also the version check as written will never fail I think.
